In [6]:
!pip install  scikit-learn


  Using cached threadpoolctl-3.6.0-py3-none-any.whl.metadata (13 kB)
   ---------------------------------------- 0.0/8.7 MB ? eta -:--:--
   --- ------------------------------------ 0.8/8.7 MB 5.6 MB/s eta 0:00:02
   ------------ --------------------------- 2.6/8.7 MB 7.6 MB/s eta 0:00:01
   -------------- ------------------------- 3.1/8.7 MB 7.7 MB/s eta 0:00:01
   ---------------------- ----------------- 5.0/8.7 MB 6.7 MB/s eta 0:00:01
   -------------------------------- ------- 7.1/8.7 MB 7.2 MB/s eta 0:00:01
   ---------------------------------------- 8.7/8.7 MB 7.4 MB/s eta 0:00:00
   ---------------------------------------- 0.0/38.6 MB ? eta -:--:--
   -- ------------------------------------- 2.4/38.6 MB 12.2 MB/s eta 0:00:03
   ---- ----------------------------------- 4.5/38.6 MB 11.7 MB/s eta 0:00:03
   ------ --------------------------------- 6.3/38.6 MB 11.3 MB/s eta 0:00:03
   ------- -------------------------------- 6.8/38.6 MB 10.0 MB/s eta 0:00:04
   -------- ------------


[notice] A new release of pip is available: 25.0.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [10]:
# === Импорт библиотек ===
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error, r2_score

# === Загрузка данных ===
data = pd.read_csv("C:\\Users\\Aleks\\Downloads\\train_data.csv")



# Просмотр первых строк для проверки структуры
display(data.head())
print(len(data['region'].unique()))

,id,atm_group,address_raw,address_geocoded,geo_lon,geo_lat,country,region,municipality,city,...,nearest_public_transport_dist_m,count_public_transport_300m,nearest_parking_dist_m,count_parking_300m,nearest_education_dist_m,count_education_300m,nearest_subway_dist_m,nearest_post_offices_dist_m,count_post_offices_300m,has_subway_nearby
0,5.0,496.5,BUDENNOGO 7A ELISTA,"Россия, Республика Калмыкия, Элиста, улица С.М...",44.260605,46.318231,Россия,Республика Калмыкия,городской округ Элиста,Элиста,...,93.6,5,143.7,3,247.3,1,0.0,0.0,0,False
1,6.0,496.5,"HO CHI MIHN AVE, 19 ULYANOVSK","Россия, Ульяновск, проспект Хо Ши Мина, 19",48.300652,54.270443,Россия,Ульяновская область,городской округ Ульяновск,Ульяновск,...,89.9,6,NaN,0,260.8,1,0.0,220.4,2,False
2,7.0,496.5,SHELESTA 116A KHABAROVSK,"Россия, Хабаровск, улица Шелеста, 116А",135.052594,48.520497,Россия,Хабаровский край,городской округ Хабаровск,Хабаровск,...,33.8,8,112.9,6,0.0,0,0.0,186.6,2,False
3,8.0,496.5,ORDZHONIKIDZE 52 YAKUTSK,"Россия, Республика Саха (Якутия), Якутск, улиц...",129.721308,62.025566,Россия,Республика Саха (Якутия),городской округ Якутск,Якутск,...,119.8,7,246.9,4,195.4,5,0.0,167.2,1,False
4,10.0,496.5,"VETERANOV AVE, 3 KRASNOKAMENS","Россия, Забайкальский край, Краснокаменск, про...",118.027480,50.090714,Россия,Забайкальский край,Краснокаменский муниципальный округ,Краснокаменск,...,70.6,3,48.3,5,NaN,0,NaN,NaN,0,False


76


In [11]:
# === Разделение данных на признаки и целевую переменную ===
# === Удаляем строки с пропущенными значениями ===
data = data.dropna()

# === Удаляем координаты ===
X = data.drop(columns=['target', 'geo_lon', 'geo_lat'])
y = data['target']

# === Кодируем категориальный признак region ===
X = pd.get_dummies(X, columns=['region'], drop_first=True)

# === Теперь все признаки числовые ===
X = X.select_dtypes(include=['int64', 'float64', 'uint8'])

# === Базовый сплит данных (25% тест, seed = 42) ===
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42
)


In [12]:
# === Обучение ===
model = Ridge(alpha=1.0, solver='auto', max_iter=1000)
model.fit(X_train, y_train)

# === Предсказание ===
y_pred = model.predict(X_test)

# === Метрики ===
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print(f"RMSE: {rmse:.4f}")
print(f"R²: {r2:.4f}")

# === Коэффициенты модели ===
print("\nКоэффициенты:", model.coef_)
print("Свободный член (intercept):", model.intercept_)

RMSE: 0.0635
R²: 0.4116

Коэффициенты: [-1.54031387e-05  2.50419294e-05  5.28133758e-06 -1.11359717e-05
  2.84318591e-03  1.74874718e-05 -1.49411348e-03  2.11797274e-05
  1.05386925e-03 -1.12365576e-04 -1.14128032e-05 -5.99187187e-04
 -2.72411388e-06  7.25129550e-04 -5.94154297e-06  1.32553147e-04
  2.34472966e-05  3.55046642e-04 -2.37820491e-05 -1.53758717e-03
 -6.45467105e-06  1.45567562e-05 -2.11106090e-03]
Свободный член (intercept): -0.04888876437929235


In [13]:
from sklearn.linear_model import RidgeCV
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# === Список возможных значений α ===
alphas = [0.001, 0.01, 0.1, 1.0, 10.0, 100.0, 1000.0]

# === RidgeCV автоматически выберет лучшее значение α ===
model = RidgeCV(alphas=alphas, cv=5)  # store_cv_values убираем
model.fit(X_train, y_train)

# === Предсказание ===
y_pred = model.predict(X_test)

# === Метрики ===
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print(f"Лучшее значение alpha: {model.alpha_:.3f}")
print(f"RMSE: {rmse:.4f}")
print(f"R²: {r2:.4f}")


Лучшее значение alpha: 1000.000
RMSE: 0.0634
R²: 0.4124


### Вывод и модификации данных

Для повышения интерпретируемости и устойчивости модели были выполнены следующие этапы предобработки данных:

1. **Удаление пропущенных значений**  
   Из исходного набора данных удалены строки с отсутствующими значениями (`NaN`), чтобы обеспечить корректную работу алгоритма Ridge-регрессии и исключить влияние некорректных наблюдений.

2. **Исключение координат (`geo_lon`, `geo_lat`)**  
   Географические координаты не несут прямой линейной зависимости с целевой переменной и могут вносить шум в модель.  
   Так как Ridge-регрессия строит линейную аппроксимацию, использование координат может искажать распределение весов и снижать обобщающую способность модели.  
   В связи с этим координаты были удалены.

3. **Кодирование категориального признака `region`**  
   Признак `region` содержал 76 уникальных значений, соответствующих регионам размещения банкоматов.  
   Для его включения в модель был применён **One-Hot Encoding** через функцию `pd.get_dummies(columns=['region'], drop_first=True)`.  
   Таким образом, каждый регион был представлен бинарным столбцом (0/1), что позволило сохранить региональные различия без нарушения линейности.  
   Использование параметра `drop_first=True` предотвратило проблему мультиколлинеарности.

4. **Удаление прочих категориальных признаков**  
   После кодирования `region` в модели остались только числовые признаки, включая сгенерированные бинарные индикаторы для каждого региона.

---

### Обучение модели Ridge-регрессии

Для обучения использовалась **Ridge-регрессия** — линейная модель с \( L_2 \)-регуляризацией, минимизирующая функционал:

$$
J(w) = \mathrm{MSE} + \alpha \lVert w \rVert^2
$$

Оптимальное значение коэффициента регуляризации подбиралось методом кросс-валидации через `RidgeCV`.  
Результаты подбора показали, что оптимальным значением является:

$$
\alpha = 1000
$$

---

### Результаты обучения и тестирования

**Базовая модель Ridge (α = 1.0):**
- RMSE: 0.0635  
- R²: 0.4116  

**Оптимизированная модель RidgeCV (α = 1000):**
- RMSE: 0.0634  
- R²: 0.4124  

---

### Интерпретация результатов

- После удаления координат и добавления кодированных регионов качество модели **осталось стабильным**, при этом удалось повысить интерпретируемость коэффициентов и снизить влияние мультиколлинеарных признаков.  
- Оптимизация параметра регуляризации (α = 1000) привела к **незначительному улучшению метрики R²** и более сглаженному распределению весов, что говорит о снижении переобучения.  
- Учитывая невысокое значение R² (~0.41), модель описывает часть дисперсии целевой переменной, однако дальнейшее улучшение возможно за счёт добавления содержательных признаков.  
- Закодированный признак `region` позволил учесть влияние территориальных различий без использования координат, что соответствует предпосылкам линейной регрессии и улучшает устойчивость модели при обобщении.

Таким образом, проведённые модификации данных (удаление координат и кодирование `region`) обеспечили корректное применение Ridge-регрессии и повысили стабильность полученной модели.
